### DISCUSSION TITLE
[d3-array] bin() defaults to HistogramGeneratorNumber<number, number> and locks out Datum

### DISCUSSION THREAD

**Discussion by gordogato81** [DefinitelyTyped/DefinitelyTyped#59496] [Issues with a @types package]
### Expected Behavior:
Calling `d3.bin()` should automatically detect whether the data type it is being given is of type <number, number>, <Datum, number>, or <Datum, Date>.

### Actual Behavior: 
`d3.bin()` defaults to interface `HistogramGeneratorNumber<number, number>` and incorrectly types inputs of type <Datum, number> and <Datum, Date> requiring manual casting.

### The problem child:
The line `export function bin(): HistogramGeneratorNumber<number, number>;` in index.d.ts appears to be the troublemaker.
If it is not possible to automatically type this function, it should at least be mentioned in its description.

[a comment by an automated account is not shown]

**Reply by gustavderdrache** [reply]
TypeScript can't automatically infer the return type, so the default signature of `d3.bin()` is a convenience to match D3's internal defaults.

If you're using a non-default `Datum` type parameter, you'll need to provide it explicitly to match the other overloads:

https://github.com/DefinitelyTyped/DefinitelyTyped/blob/a0754cf587d5378dca53e0ebb905439d0e94698a/types/d3-array/v2/index.d.ts#L1098-L1100

Something like this should suffice:

```ts
const bin = d3.bin<Datum, Date>();
```

**Reply by gordogato81** [reply]
I did find that out, but only by looking at the typings/d3-array/index.d.ts file. This typing requirement is not documented anywhere that I could find, which I think is a bit of a problem. Maybe it would be possible to add this information in the function description?
